app/models/gem_download.rb

```diff
@@ -4,89 +4,121 @@ class GemDownload < ActiveRecord::Base
+      # insert = "INSERT INTO #{quoted_table_name} (rubygem_id, version_id, count) SELECT ?, ?, ?"
+      # find_by_sql(["WITH upsert AS (#{update}) #{insert} WHERE NOT EXISTS (SELECT * FROM upsert)", count, rubygem_id, version_id, count]).first
+      find_by_sql([update, count]).first
+    end
 
-  # Takes an array where members have the form
-  #   [full_name, count]
-  # E.g.:
-  #   ['rake-10.4.2', 1]
-  def self.bulk_update(ary)
-    updates_by_version = {}
-    updates_by_gem = {}
-
-    ary.each do |full_name, count|
-      if updates_by_version.key?(full_name)
-        version, old_count = updates_by_version[full_name]
-        updates_by_version[full_name] = [version, old_count + count]
-      else
-        version = Version.find_by(full_name: full_name)
-        updates_by_version[full_name] = [version, count] if version
+    # Takes an array where members have the form
+    #   [full_name, count]
+    # E.g.:
+    #   ['rake-10.4.2', 1]
+    def bulk_update(ary)
+      updates_by_version = {}
+      updates_by_gem = {}
+
+      ary.each do |full_name, count|
+        if updates_by_version.key?(full_name)
+          version, old_count = updates_by_version[full_name]
+          updates_by_version[full_name] = [version, old_count + count]
+        else
+          version = Version.find_by(full_name: full_name)
+          updates_by_version[full_name] = [version, count] if version
+        end
       end
-    end
 
-    updates_by_version.values.each do |version, version_count|
-      updates_by_gem[version.rubygem_id] ||= 0
-      updates_by_gem[version.rubygem_id] += version_count
-    end
+      return if updates_by_version.empty?
+      updates_by_version.values.each do |version, version_count|
+        updates_by_gem[version.rubygem_id] ||= 0
+        updates_by_gem[version.rubygem_id] += version_count
+      end
+
+      updates_by_version.values.sort_by { |v, _| v.id }.each do |version, count|
+        # Gem version count
+        increment(count, rubygem_id: version.rubygem_id, version_id: version.id)
+      end
+
+      update_gem_downloads(updates_by_gem)
+
+      total_count = updates_by_gem.values.sum
 
-    updates_by_version.values.sort_by { |v, _| v.id }.each do |version, count|
-      # Gem version count
-      increment(count, rubygem_id: version.rubygem_id, version_id: version.id)
+      # Total count
+      increment(total_count, rubygem_id: 0, version_id: 0)
     end
 
-    updates_by_gem.sort_by { |id, _| id }.each do |rubygem_id, count|
-      # Gem count
-      increment(count, rubygem_id: rubygem_id, version_id: 0)
+    private
+
+    # updates the downloads field of rubygem in DB and ES index
+    # input: [rubygem_id, download_count_to_increment]
+    def update_gem_downloads(gem_downloads)
+      bulk_update_query = []
+
+      gem_ids = gem_downloads.map { |id, _| id }
+      downloads = GemDownload.where("rubygem_id IN (#{gem_ids.join(',')})").pluck(:count)
```

Comment:
`gem_ids`is array with square brackets `[id1, id2, id3]`. I want to look for all `GemDownloads` whose `count` will be updated in ES index and using `IN (array)` would get it done one query, except `array` syntax should be a `comma` separated value with no `[]`.